### PR TITLE
[FCL-343] Add typing to utils

### PR DESCRIPTION
**Pull request by jacksonj04**
We prefer our code to be typed, as it reduces opportunities for misunderstanding.

Enforce typing in utilities, and add missing typing where needed.

### CODE REVIEW

**Comment by dragon-dxw** [comment]
API client passes tests and typechecks correctly with these changes.

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -9,15 +9,53 @@ repos:
...
+        exclude: courts_schema_types_autogenerated\.py
       - id: forbid-submodules
       - id: mixed-line-ending
       - id: no-commit-to-branch
       - id: trailing-whitespace
 
+  # There is a bug with upstream which means the imports are
+  # not sorted -- this should do for now
+  # - repo: https://github.com/camptocamp/jsonschema-gentypes
+  - repo: https://github.com/dxw/jsonschema-gentypes
```
Comment (jacksonj04) [inline .pre-commit-config.yaml:21]: This has now been patched upstream, so no need to use our own fork.

File: .pre-commit-config.yaml [hunk 2]
```diff
@@ -9,15 +9,53 @@ repos:
       - id: check-xml
       - id: check-yaml
       - id: end-of-file-fixer
+        exclude: courts_schema_types_autogenerated\.py
```
Comment (jacksonj04) [inline .pre-commit-config.yaml:12]: Can we just exclude this across the board by adding it to the top of the config? Seems daft to add it in all the stages when we know it's autogenerated and will probably upset a lot of quality tools.

File: jsonschema-gentypes.yaml [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+    root_name: RawCourtRepository
+    # Argument passed to the API
+    api_arguments:
+      additional_properties: Only explicit
+    # Rename an element
+    name_mapping: {}
+    # The minimum Python version that the code should support. By default the
+    # currently executing Python version is chosen. Note that the output
+    # may require typing_extensions to be installed.
+    python_version: "3.9"
```
Comment (jacksonj04) [inline jsonschema-gentypes.yaml:22]: Weird that we need this in two places.

File: pyproject.toml [hunk 4]
```diff
@@ -49,3 +39,6 @@ update_changelog_on_bump = true
 [build-system]
 requires = ["poetry-core>=1.0.0"]
 build-backend = "poetry.core.masonry.api"
+
+[tool.mypy]
+exclude = ["ds_caselaw_utils/courts_schema_types_autogenerated.py"]
```
Comment (jacksonj04) [inline pyproject.toml:44]: If we globally exclude this in pre-commit config, do we need this line?

File: src/ds_caselaw_utils/courts.py [hunk 5]
```diff
@@ -1,97 +1,119 @@
+# mypy: disable-error-code="override"
```
Comment (jacksonj04) [inline src/ds_caselaw_utils/courts.py:1]: Not a huge fan of this, but happy to leave and fix later if needed. Might be worth seeing if we've inadvertently fixed it in other tidying and can remove it though.

File: src/ds_caselaw_utils/courts.py [hunk 6]
```diff
@@ -1,97 +1,119 @@
...
+from typing import TYPE_CHECKING, NewType, Optional
+
+if TYPE_CHECKING:
+    from courts_schema_types_autogenerated import (  # type: ignore[import-not-found]
+        RawCourt,
+        RawCourtRepositoryData,
+        RawJurisdiction,
+    )
+else:
+    RawCourt = RawCourtRepositoryData = RawJurisdiction = None
```
Comment (jacksonj04) [inline src/ds_caselaw_utils/courts.py:18]: Do we still need this workaround?

**Review by dragon-dxw — APPROVED** [review]

File: jsonschema-gentypes.yaml (hunk 3 above)
Comment (dragon-dxw) [reply jsonschema-gentypes.yaml:22]: It's probably a bug that it didn't respect this value -- I suspect this should be inverted, so that it treats the python version as the minimum of current and minimum permissable.

https://github.com/sbrunner/jsonschema-gentypes/blob/58d7b7ee112ca20fe77726b42b9594c0eb323f97/jsonschema_gentypes/__init__.py#L374

File: src/ds_caselaw_utils/courts.py (hunk 5 above)
Comment (dragon-dxw) [reply src/ds_caselaw_utils/courts.py:1]: It's not removable; it's essentially caused by the attributes being theoretically settable (they're just generic attributes), but the subclass doesn't have setters (because we never change them). There doesn't seem to be a clean way of making those attributes non-settable...

File: pyproject.toml (hunk 4 above)
Comment (dragon-dxw) [reply pyproject.toml]: Needed a lot of [necessary] tidying to remove it, but yes

File: src/ds_caselaw_utils/courts.py (hunk 5 above)
Comment (jacksonj04) [reply src/ds_caselaw_utils/courts.py:1]: This is probably into the realms of the abstract base class stuff, which we _should_ do at some point but for now isn't important so happy to leave this.

File: jsonschema-gentypes.yaml (hunk 3 above)
Comment (jacksonj04) [reply jsonschema-gentypes.yaml:22]: Agree the logic is hard to follow there. Suspect you're right, but either way happy to just peg this.
